Introduce new "smart" range-based team building options

jdoerfert opened the pull request:
The old team building options where either fixed (1,2,3,4) or purely
random (?). The new options (1-2,1-3,1-4) allow players to be placed
in ranges of teams while they also ensure team sizes to be as even as
possible. That is, a player with the team `1-3` is placed in team 1,
2, or 3 in a way that minimizes the absolute difference in team size.

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -21,17 +21,20 @@
 #include "Settings.h"
 #include "commonDefines.h"
 #include "files.h"
+#include "helpers/MultiArray.h"
```
```suggestion
```
Unused?

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+                nPotentialPlayers[3].insert(player);
+                break;
+            case TM_NOTEAM: break;
+            default: RTTR_Assert_Msg(false, "Unexpected team!");
```
What about the random teams?

jdoerfert replied:
They don't "exist" this late. Whenever the host clicks "random" they are assigned, for now, to a team (IIRC). A follow up could change that and distribute random teams here as well, maybe make ? behave like 1-4 does and remove the latter. That would be a breaking change and I wanted to separate things so people can determine if they want to keep the old behavior at all or simply go with "smart" teams.

Flamefire replied:
IIRC when "random" is selected a TM_RANDOMTEAM1-TM_RANDOMTEAM4 is assigned which shows up as random but is a separate enum. Should be easy enough to test by trying it.

jdoerfert replied:
My bad, they are assigned early but keep the TM_RANDOMTEAMX value still. I'll add them here and keep the old behavior for now.

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+    if(unassignedPlayers == 0)
+        return false;
+
+    RTTR_Assert_Msg(!nPotentialPlayers[0].empty() || !nPotentialPlayers[1].empty() || !nPotentialPlayers[2].empty()
```
I'd remove that due to size of the condition. The same issue will be caught a bit below by the next assertion(s)

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+        std::vector<int> teamsForNextPlayer;
+
+        // Determine the minimal team size for teams that can take an unassigned player.
+        unsigned minNextTeamSize = -1;
```
This is a bug, I'd suggest to write tests to catch such issues.

jdoerfert replied:
I don't get it. What is a bug? The assertions should never trigger, as the code is written. They are safeguards for future changes that break invariants. Unsure what you want to test and catch here.

Flamefire replied:
Converting `-1` to unsigned makes it a large value. I got confused by the `std::min`, so that is not a bug, but confusing to read. You can use `std::numeric_limits<unsigned>::max()` instead or use a boost::optional

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+            if(!nPotentialPlayers[team].empty())
+              minNextTeamSize = std::min(minNextTeamSize, nPlayers[team]);
+        }
+        RTTR_Assert_Msg(minNextTeamSize != -1, "Expected to find team for an unassigned player!");
```
This can be removed if you initialize `minNextTeamSize=0`

jdoerfert replied:
But that would simply hide a broken invariant, right? The point was to ensure we do not end up here without having potential players to distribute, though, I can obviously remove the asserts if that is preferred.

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+            teamsForNextPlayer.push_back(team);
+        }
+
+        int nextTeam = teamsForNextPlayer[rand() % teamsForNextPlayer.size()];
```
An assert here makes sense: teamsForNextPlayer must not be empty. In meaning that assert is the same as the one above (line 499)

jdoerfert replied:
Sure. I'll add one here, remove the others.

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+        int nextPlayer = *it;
+
+        // The player is now assigned and the team size increased.
+        for(int i = 0; i < 4; ++i)
+            nPotentialPlayers[i].erase(nextPlayer);
```
Maybe:
```suggestion
        for(auto& playerSet: nPotentialPlayers)
            playerSet.erase(nextPlayer);
```

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +434,112 @@ void GameServer::Stop()
+// number of players per team. Returns true iff players have been assigned.
+static bool adjustSmartTeams(std::vector<JoinPlayerInfo>& playerInfos)
+{
+    std::set<unsigned> nPotentialPlayers[4];
```
Could this get a better name? It isn't the number of players but player indices so remove the "n" prefix?

jdoerfert replied:
sure.

Flamefire reviewed (changes requested):
I do like the idea. While you are at it: I think it makes sense to overhaul the whole random team thing so we have either
- Team 1, 2, 3 or 4
- Random
- Fair Random with restriction (yours)

We currently "fake" the Random thing by using different enum values. Having your code already let's us do that properly in a similar way: Just iterate over the "Random" values and assign a random team, then broadcast.
This could be done here or in a follow-up PR.

For this PR I'd like to see some tests. If you make that function non-static it should be pretty easy to test: Throw in some players with teams set and check that the distribution of teams matches (i.e. no overfull team or a team with no players) and no player is in a team where he shouldn't be (when using the restriction to team 1-3 etc). For this I suggest a helper function which creates a vector of JoinPlayerInfo out of a vector of Teams or even fully wrap it in a (test) function which takes a vector of teams, calls your function and returns a vector of teams. That is then easily testable :)

jdoerfert commented:
> For this PR I'd like to see some tests. If you make that function non-static it should be pretty easy to test: Throw in some players with teams set and check that the distribution of teams matches (i.e. no overfull team or a team with no players) and no player is in a team where he shouldn't be (when using the restriction to team 1-3 etc). For this I suggest a helper function which creates a vector of JoinPlayerInfo out of a vector of Teams or even fully wrap it in a (test) function which takes a vector of teams, calls your function and returns a vector of teams. That is then easily testable :)

I could not find any tests referencing joint players or RANDOMTEAM, where should they be placed?

Spikeone commented:
Dropdowns for team (and color) selection would be great as well then :) less clicking for everyone

Flamefire commented:
This folder is probably the right place: https://github.com/Return-To-The-Roots/s25client/tree/master/tests/s25Main/integration
You can use https://github.com/Return-To-The-Roots/s25client/blob/master/tests/s25Main/integration/testGameServerPlayer.cpp as a starting point and ping me for questions

jdoerfert commented:
Looking at the CI error, I'm unsure I caused this.

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,136 @@
+#include "JoinPlayerInfo.h"
+#include "network/GameServer.h"
+#include "gameTypes/TeamTypes.h"
+#include <boost/test/tools/old/interface.hpp>
```
Not required
```suggestion
```

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,136 @@
+#include "gameTypes/TeamTypes.h"
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+#include <cstdio>
```
Not required(?)
```suggestion
```

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    std::sort(&nPlayers[0], &nPlayers[nTeams]);
+    BOOST_REQUIRE_LE(nPlayers[0], nPlayers[nTeams - 1]);
```
I found it easier to switch to the new Boost test macros:
```suggestion
    BOOST_TEST_REQUIRE(nPlayers[0] < nPlayers[nTeams - 1]);
```

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,136 @@
+    playerInfos.emplace_back(BPIRT2);
+    playerInfos.emplace_back(BPI1_3);
+    BOOST_REQUIRE(assignPlayersOfRandomTeams());
+    BOOST_REQUIRE_GE(2, getMaxTeamSizeDifference(playerInfos, 4));
```
This is the source of the CI error (unsigned 2):

```suggestion
    BOOST_TEST(2u >= getMaxTeamSizeDifference(playerInfos, 4));
```

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,136 @@
+    auto& playerInfos = getPlayerInfos();
+
+    playerInfos.clear();
+    playerInfos.emplace_back(BPI1_4);
```
No need for emplace, push_back is enough for all

jdoerfert replied:
Any particular reason? I mean, I don't care much but it seems rather counter intuitive.

Flamefire reviewed (changes requested):
Hm, in the previous iteration you had a static method which received a vector of players. That would have been easier to test as you wouldn't even need the fixture to create a GameServer instance. What was the reason for switching that?
In general such pure functions are preferable as they don't require any setup for tests and are easier to reason about as they don't have state

jdoerfert commented:
> Hm, in the previous iteration you had a static method which received a vector of players. That would have been easier to test as you wouldn't even need the fixture to create a GameServer instance. What was the reason for switching that?
> In general such pure functions are preferable as they don't require any setup for tests and are easier to reason about as they don't have state

As long as it is static (=internal linkage), that all makes sense. Adding global externally visible functions is usually not what people advocate for, but, y'know, I can go with a global one and pass a vector instead.

jdoerfert commented:
> Looking at the CI error, I'm unsure I caused this.

^ still the only failure

jdoerfert commented:
ping

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +433,111 @@ void GameServer::Stop()
+        RTTR_Assert_Msg(!PotentialPlayers[nextTeam].empty(), "Expected next team to have potential players!");
+
+        // Pick a random player that can go into this team.
+        auto it = PotentialPlayers[nextTeam].begin();
+        std::advance(it, rand() % PotentialPlayers[nextTeam].size());
+        int nextPlayer = *it;
```
I feel this should go into a new function. This is rather verbose and hence hard to see what it does, and used a couple of times. So my suggestion: Create a function `element getRandomElement(collection, prng)` with another overload without the prng (which creates one internally and calls the other function)
Or better: Instead of a prng it could take a functor `getRandIndex(max)` which returns a value in `[0,max]`. We have several other uses in the code where we need a random element from a container, so that would certainly help.

Can be done in a follow-up though.

Flamefire commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +433,111 @@ void GameServer::Stop()
+// number of players per team. Returns true iff players have been assigned.
+bool GameServer::assignPlayersOfRandomTeams(std::vector<JoinPlayerInfo>& playerInfos)
+{
+    std::set<unsigned> PotentialPlayers[4];
```
I know this isn't introduced by you but could you use a named constant instead of the 4? You use it quite a lot so a constant makes sense. Maybe rename the NUM_TEAMS to NUM_TEAM_OPTIONS (careful to do it in all code) and introduce a new NUM_TEAMS=4

Flamefire commented on tests/s25Main/integration/testGameServerJointPlayerDistribution.cpp:
```diff
@@ -0,0 +1,129 @@
+    playerInfos.push_back(JoinPlayerInfo(BPIRT2));
+    playerInfos.push_back(JoinPlayerInfo(BPI1_4));
+    BOOST_REQUIRE(GameServer::assignPlayersOfRandomTeams(playerInfos));
+    BOOST_REQUIRE_EQUAL(0, getMaxTeamSizeDifference(playerInfos, 4));
```
```suggestion
    BOOST_TEST(0u == getMaxTeamSizeDifference(playerInfos, 4));
```
Similar below to fix the CI issue. Notice the use of the unsigned constant (which is what MSVC complains about: currently there is "signed == unsigned" used)

Flamefire reviewed (changes requested):
Ok, next round. Getting close :+1: 

Optional: I suggest you pull from RttR master and rebase your branch on top of it (that is the note from Github):
```
git fetch origin
git rebase origin/master
```

jdoerfert commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -430,13 +433,111 @@ void GameServer::Stop()
+        JoinPlayerInfo& playerInfo = playerInfos[player];
+        switch(playerInfo.team)
+        {
+            case TM_RANDOMTEAM:
```
RANDOMTEAM is not random, it is `?` that was resolved to `1` (= team 1),
RANDOMTEAM2 is not random, it is `?` that was resolved to `2` (= team 2),
...

jdoerfert replied:
I assume this is resolved w/o change.

jdoerfert commented:
> Optional: I suggest you pull from RttR master and rebase your branch on top of it (that is the note from Github):
> 
> ```
> git fetch origin
> git rebase origin/master
> ```

I did when I created the pull request. I thought it is still mergeable, isn't it?

Flamefire commented:
> I did when I created the pull request. I thought it is still mergeable, isn't it?

Yes, hence why I marked it as optional. Github shows the "This branch is out-of-date with the base branch" notice, so there have been commits to master since then. Not sure what is inbetween now but it doesn't hurt and avoids conflicts if you make more changes. But as mentioned: Optional, feel free to ignore

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+
+/// Return a random number in the interval [lowerBound, upperBound].
+template<typename RandomT>
+static unsigned getRandomNumber(const RandomT& rng, unsigned lowerBound, unsigned upperBound)
```
```suggestion
inline unsigned getRandomNumber(RandomT& rng, unsigned lowerBound, unsigned upperBound)
```

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+template<typename RandomT>
+static unsigned getRandomNumber(const RandomT& rng, unsigned lowerBound, unsigned upperBound)
+{
+    std::uniform_int_distribution<typename RandomT::result_type> dist(lowerBound, upperBound);
```
The result of the PRNG doesn't matter for the result if the distribution:
```suggestion
    std::uniform_int_distribution<unsigned> dist(lowerBound, upperBound);
```

jdoerfert replied:
Why?

Flamefire replied:
You want an unsigned number that is uniformly distributed, don't you? The change makes it say that. Internally it could call the rng multiple times to get enough entropy to fill an unsigned. E.g. the rng could return only bytes.

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+
+/// Return a random number in the interval [0, upperBound).
+template<typename RandomT>
+static unsigned getRandomNumber(const RandomT& rng, unsigned upperBound)
+{
+    RTTR_Assert(upperBound > 0);
+    return getRandomNumber(rng, 0, upperBound - 1);
```
Better: 
```suggestion
inline unsigned getRandomIndex(const RandomT& rng, unsigned size)
{
    RTTR_Assert(size > 0);
    return getRandomNumber(rng, 0, size - 1);
```

jdoerfert replied:
> See comments. Mostly about inline and non-const rng.
> Also a suggestion about not naming a non-inclusive range "upperBound". I think this makes it clearer

- const rng was an oversight.
- inline seems to be counter productive. What is the reason you want it to have inline linkage over internal linkage?

Flamefire replied:
Inline mainly because it is in a header. If it was a cpp file, then yes, internal linkage but this way it would be internal to any cpp file which includes this header which isn't what one usually wants. Besides that the compiler warns about unused static functions but not for inline functions, so that's what I'd use

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+}
+
+/// Return a random number in the interval [0, upperBound).
+static unsigned getRandomNumber(unsigned upperBound)
+{
+    std::mt19937 rng;
+    std::random_device rnd_dev;
+    rng.seed(rnd_dev());
+    RTTR_Assert(upperBound > 0);
+    return getRandomNumber(rng, upperBound);
```
```suggestion
inline unsigned getRandomIndex(unsigned size)
{
    std::random_device rnd_dev;
    std::mt19937 rng(rnd_dev());
    return getRandomNumber(rng, size);
```

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+/// Return an iterator to a random element of \p container. If \p isEnd is nonnull, set it to true if the returned
+/// iterator is the "end" iterator and false otherwise.
+template<typename ContainerT>
+static typename ContainerT::iterator getRandomElement(ContainerT& container, bool* isEnd = nullptr)
```
```suggestion
inline typename ContainerT::iterator getRandomElement(ContainerT& container, bool* isEnd = nullptr)
```

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+    if(empty)
+        return container.end();
+    auto it = container.begin();
+    std::advance(it, getRandomNumber(size));
```
```suggestion
    std::advance(it, getRandomIndex(size));
```

Flamefire commented on libs/s25main/random/Random.h:
```diff
@@ -139,3 +140,45 @@ class RandomFunctor
+
+/// Return a random number in the interval [0, upperBound).
+template<typename RandomT>
+static unsigned getRandomNumber(const RandomT& rng, unsigned upperBound)
```
```suggestion
inline unsigned getRandomNumber(RandomT& rng, unsigned upperBound)
```

Flamefire reviewed (changes requested):
See comments. Mostly about inline and non-const rng.
Also a suggestion about not naming a non-inclusive range "upperBound". I think this makes it clearer

Flamefire commented:
Hey @jdoerfert any progress here?

jdoerfert commented:
> Hey @jdoerfert any progress here?

No. I might revisit this at some point but have no intention of doing so in the near future. Feel free to use the code and adapt it yourself.

FWIW, the review process, especially given the state of the code base, should be reconsidered. I get it that you want to make things perfect, for some definition of perfect, but that is not how code reviews will encourage participation and long term engagement. Just my 2c.

Flamefire commented:
> FWIW, the review process, especially given the state of the code base, should be reconsidered. I get it that you want to make things perfect, for some definition of perfect, but that is not how code reviews will encourage participation and long term engagement. Just my 2c.

Point well taken. I'm working hard (every once in a while at least) to get the code base to a better state so it is also easier to maintain.
Any specific points of the review you'd have suggestions on how to do better?
I tried to provide suggestions for the change so it would be low effort from your side to use them (click 1 button) and about 80% of the comments were suggestions on how to make the code compile/work/pass CI. A few nits were with naming, e.g. variables as lowercase (we should write that down somewhere I guess) and not introducing new magic constants (one of the things I'm removing over time where they still exist)

Anyway, thanks for this very useful feature! I'll take care of the remaining few issues and get this in.

jdoerfert commented:
> > FWIW, the review process, especially given the state of the code base, should be reconsidered. I get it that you want to make things perfect, for some definition of perfect, but that is not how code reviews will encourage participation and long term engagement. Just my 2c.
> 
> Point well taken. I'm working hard (every once in a while at least) to get the code base to a better state so it is also easier to maintain.

Don't get me wrong, I did not want to argue "against" the code base. I wanted to argue for "imperfect" additions that can be refined in-tree.

> Any specific points of the review you'd have suggestions on how to do better?

From what I've read on this topic, I decided to accept "correct" and "reasonable" code rather earlier than later. More patches and quicker turn around time is more encouraging than bugged down on a feature till it's "perfect", it never is anyway. There can be rules, e.g.,requiring tests is fine, but non-rule based change request should be optional and limited to a minimum. For example, if something like assertions or comments are supposed to be "too much", don't require a new round of review just to remove them.  Making suggestions is generally OK, but you should aim at the least amount of review rounds. One more option is a conditional accept. If changes are necessary before it can be merged, provide details so the next version is OK. Code can be improved in-tree as well, possibly faster. Maybe even ask the author to review those follow up patches so you even get the benefit of showing how you would the code to look like.

> I tried to provide suggestions for the change so it would be low effort from your side to use them (click 1 button) and about 80% of the comments were suggestions on how to make the code compile/work/pass CI. A few nits were with naming, e.g. variables as lowercase (we should write that down somewhere I guess) and not introducing new magic constants (one of the things I'm removing over time where they still exist)

I'm not very familiar with the github API, maybe that is on me. The problem with nits is that they should not require more review rounds. People need to make time to look at comments, read them, go to the repo, modify the files, upload, reply, go back to "wait mode", which is not perfect, especially in the beginning. 

> Anyway, thanks for this very useful feature! I'll take care of the remaining few issues and get this in.

Thanks, that would be great.

---

I'll check at some point if I have more features that might be upstreamable. For now, I'll concentrate on my bot which needs a redesign to keep timing issues in check.

Flamefire commented:
>  I wanted to argue for "imperfect" additions that can be refined in-tree.

Experience has shown that once code is accepted it often won't be "polished"/refined. It is *much* easier to do that in the review. So some level of quality must be achieved prior to merging. Most of that is guarded by CI already, i.e. warnings and unit tests.
Especially the latter are crucial for any new feature and useful for bugfixes.

> For example, if something like assertions or comments are supposed to be "too much", don't require a new round of review just to remove them. Making suggestions is generally OK, but you should aim at the least amount of review rounds

I think I mentioned those together with actual issues, i.e. compile failures due to (at least) warnings. But alright, I will be more careful here, thanks!

> Code can be improved in-tree as well, possibly faster.

I'd disagree here. The work is the same: You commit, push, done. The benefit doing that in one PR is that you don't need to open a new PR (which is some work) and that the initial PR shows the whole changes in one diff. So you can easily see what belongs together and why things are introduced/changed. In follow-ups this is harder as you likely need to open the whole code base to see where the changes comes into play.

> I'm not very familiar with the github API, maybe that is on me. 

In case that helps: The things mentioned show up as "Suggested change". You can (I think at least when in the "files" tab) click the "commit" button below it to accept that change and commit it and even add multiple such to a single commit (see that button). You might then want to locally pull and see if any surrounding code needs updating too but often that button is enough. It is a kinda new feature and I found that useful when having my code reviewed.

Flow86 commented:
> Code can be improved in-tree as well, possibly faster.
I also have to disagree here. Changes which belong to each other should belong to each other - specially if one already saw them in a review.

jdoerfert commented:
> > Code can be improved in-tree as well, possibly faster.
> 
> I also have to disagree here. Changes which belong to each other should belong to each other - specially if one already saw them in a review.

I have no intention to fight you on this or start a discussion, I tried to provide you with some "best practices" I apply in other projects. You run this one as you wish.


> > I wanted to argue for "imperfect" additions that can be refined in-tree.
> 
> Experience has shown that once code is accepted it often won't be "polished"/refined. It is _much_ easier to do that in the review. So some level of quality must be achieved prior to merging. Most of that is guarded by CI already, i.e. warnings and unit 
tests.

I did not argue to include broken code, as defined by the CI for example. Imperfect could just mean that a new helper function is static and hidden instead of exposed in a new public header and then generalized, renamed, moved to a new place, and finally used in other places as well. All of that should be a follow up.

> Especially the latter are crucial for any new feature and useful for bugfixes.

Never objected to tests or working CI.

> > Code can be improved in-tree as well, possibly faster.
> 
> I'd disagree here. The work is the same: You commit, push, done. The benefit doing that in one PR is that you don't need to open a new PR (which is some work) and that the initial PR shows the whole changes in one diff. So you can easily see what belongs together and why things are introduced/changed. In follow-ups this is harder as you likely need to open the whole code base to see where the changes comes into play.

As mentioned above, your project, you do you. I guess you have more experience with the way people contribute here anyway, my feedback described what we do elsewhere but it might not be applicable.

Flamefire commented:
Follow-up in #1331, so closing this. Thanks a lot!
